common/beerocks/bcl/include/bcl/network/socket_refactorings.h

```diff
@@ -0,0 +1,795 @@
+    virtual const struct msghdr *msghdr() const = 0;
+
+    struct msghdr *msghdr()
+    {
+        return const_cast<struct msghdr *>(const_cast<const MsgHdr *>(this)->msghdr());
+    }
+};
+
+/**
+ * Any object implementing the Socket interface is a FileDescriptor too.
+ * This way we give meaning to the action of "closing the socket".
+ */
+class Socket : public FileDescriptor {
+    virtual int setsockopt(int level, int optname, void *optval, socklen_t optlen) = 0;
+};
+
+/**
+ * Abstract base class for all types of sockets: Raw, UDP, TCP, UDS, ...
+ * This implementation class aggregates a FileDescriptor implementation so it has a file
+ * descriptor. Methods overridden from FileDescriptor interface delegate on the aggregated
+ * implementation.
+ * Derived classes provide the file descriptor obtained with a call to socket(), using different
+ * family, type and protocol parameters.
+ * This class aggregates a FileDescriptor instead of inheriting from one of its implementations to
+ * follow the principle of "Favor Aggregation over Inheritance".
+ * See https://wiki.c2.com/?UseCompositionAndInterfacesWithoutClassInheritance
+ */
+class SocketAbstractImpl : public Socket {
+public:
+    int fd() override { return m_descriptor->fd(); }
+    bool close() override { return m_descriptor->close(); }
+    int setsockopt(int level, int optname, void *optval, socklen_t optlen) override
+    {
+        return ::setsockopt(m_descriptor->fd(), level, optname, optval, optlen);
+    }
+
+protected:
+    /**
+     * Constructor is protected so only derived classes can call it
+     */
+    explicit SocketAbstractImpl(int fd) : m_descriptor(std::make_shared<FileDescriptorImpl>(fd)) {}
+
+private:
+    std::shared_ptr<FileDescriptor> m_descriptor;
+};
+
+/**
+ * Classes implementing this interface model either the socket connection established at the
+ * server side when accept() system call is called or at the client side when connect() is called.
+ *
+ * The interface defines the methods to send data over a socket and to receive data from a socket.
+ */
+class Connection {
+public:
+    virtual ~Connection() = default;
+
+    /**
+     * @brief Returns the underlying socket used by this connection.
+     *
+     * Access to the underlying socket is required to obtain the socket file descriptor with which
+     * wait for read or write events using select() or epoll() functions.
+     *
+     * Connection can be closed by closing the underlying socket.
+     *
+     * @return Socket used by the connection
+     */
+    virtual std::shared_ptr<Socket> socket() = 0;
+
+    virtual int receive(Buffer &buffer)                                               = 0;
+    virtual int receive_from(Buffer &buffer, SockAddr &address)                       = 0;
+    virtual int receive_message(MsgHdr &message)                                      = 0;
+    virtual int send(const Buffer &buffer, size_t length)                             = 0;

```

Comment:
`Buffer` must encapsulate its size internally, leaving the error prone of supplying two parameters that not necessarily correlate to each other. Moreover, in current `Buffer` implementation there is a `length` property, so the confusion is even bigger.
```suggestion
    virtual int send(const Buffer &buffer)                             = 0;
```
Note:
If the intent is to **use this buffer but only to this length**, something else should be done. 
Examples:
* `send_n(const Buffer &buffer, size_t n) = 0;` //(to mimic `std::copy_n` and alike)
* Add `Buffer::limit_size(uint16_t use_only_up_to_this_byte);` // and when buffer uses an interface that needs to know the size, use this value

Reply:
Yes, the intent is to use the buffer up to given length. I chose `send` as the name of the method because it wraps around the `::send()` system call. Documentation, once written, must clearly state the purpose of the method and the meaning of all parameters.

Reply:
I don't think documentation alone will suffice, please stick to normal name conventions.
But more important, I think that if this is the case I would still encapsulate the data in the `Buffer` itself, and not externally `Buffer::limit_size(uint16_t use_only_up_to_this_byte);` with `send(Buffer&)`. because:
* it has less errors. e.g. you request more bytes than there are in the buffer.
* it is less confusing for the caller.
* it is kept along few calls. consider:
```cpp
Buffer<2000> buf;
buf.limit_size(100);
send(buf);  // first use
LOG(DEBUG) << buf;   // second use
```
compared to:
```cpp
send_n(buf,100);
LOG(DEBUG) << buf << " but using only:" << 1000; // inevitable numerous errors
```
However, having said that, if you consider using _different_ sizes in parallel, then you can't embed the limit inside the Buffer:
```cpp
send_to(buf, addr1, 100);
send_to(buf, addr2, 2000);
```